File: src/common/template_structure.py
"""Shared structural validation for imported component tags.

Both Blade and JavaScript targets resolve imported tags independently.  A
malformed source must therefore fail before either target rewrites the tags;
otherwise Blade can preserve a dangling close tag while the JS AST silently
ignores it, producing different DOM trees.
"""

from __future__ import annotations

import re


class TemplateStructureError(ValueError):
    """Raised when imported component tags are not structurally balanced."""


_TAG_NAME_RE = re.compile(r"[A-Za-z][\w-]*")
_RAWTEXT_TAGS = {"script", "style", "textarea", "title"}


def _location(source: str, index: int) -> tuple[int, int]:
    line = source.count("\n", 0, index) + 1
    line_start = source.rfind("\n", 0, index) + 1
    return line, index - line_start + 1


def _scan_tag_end(source: str, start: int) -> int:
    """Return the index after ``>`` while respecting quoted attributes."""
    quote = None
    escaped = False
    pos = start
    while pos < len(source):
        char = source[pos]
        if quote is not None:
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == quote:
                quote = None
        elif char in ("'", '"'):
            quote = char
        elif char == ">":
            return pos + 1
        pos += 1
    return len(source)


def _error(source: str, index: int, message: str) -> TemplateStructureError:
    line, column = _location(source, index)
    return TemplateStructureError(f"{message} at line {line}, column {column}.")


def validate_imported_tag_structure(source: str, imports: dict | None) -> None:
    """Validate paired/self-closing imported component tags with one stack.

    Native HTML is deliberately left to the HTML/AST pipeline.  Imported tags
    are strict because they become component lifecycle boundaries and must be
    rewritten identically for Blade and JavaScript.
    """
    imported_tags = set((imports or {}).keys())
    if not imported_tags:
        return

    stack: list[tuple[str, int]] = []
    pos = 0
    length = len(source)

    while pos < length:
        if source.startswith("{{--", pos):
            end = source.find("--}}", pos + 4)
            pos = length if end == -1 else end + 4
            continue
        if source.startswith("<!--", pos):
            end = source.find("-->", pos + 4)
            pos = length if end == -1 else end + 3
            continue
        if source[pos] != "<":
            pos += 1
            continue

        cursor = pos + 1
        is_closing = False
        if cursor < length and source[cursor] == "/":
            is_closing = True
            cursor += 1
        while cursor < length and source[cursor].isspace():
            cursor += 1

        name_match = _TAG_NAME_RE.match(source, cursor)
        if not name_match:
            pos += 1
            continue

        tag_name = name_match.group(0)
        tag_end = _scan_tag_end(source, name_match.end())
        tag_source = source[pos:tag_end]

        # Do not inspect template-like text inside native raw/RCDATA elements.
        if not is_closing and tag_name.lower() in _RAWTEXT_TAGS:
            close_match = re.search(
                rf"</\s*{re.escape(tag_name)}\s*>",
                source[tag_end:],
                flags=re.IGNORECASE,
            )
            pos = length if close_match is None else tag_end + close_match.end()
            continue

        if tag_name not in imported_tags:
            pos = tag_end
            continue

        if is_closing:
            if not stack:
                raise _error(
                    source,
                    pos,
                    f"Unexpected closing component tag </{tag_name}>; "
                    f"<{tag_name} /> is already self-closing",
                )
            open_name, open_pos = stack[-1]
            if open_name != tag_name:
                open_line, open_column = _location(source, open_pos)
                raise _error(
                    source,
                    pos,
                    f"Mismatched closing component tag </{tag_name}>; expected "
                    f"</{open_name}> for <{open_name}> opened at line "
                    f"{open_line}, column {open_column}",
                )
            stack.pop()
        elif not re.search(r"/\s*>$", tag_source):
            stack.append((tag_name, pos))

        pos = tag_end

    if stack:
        tag_name, open_pos = stack[-1]
        raise _error(
            source,
            open_pos,
            f"Unclosed component tag <{tag_name}>; expected </{tag_name}>",
        )

File: src/Template/TemplateStructure.php
<?php

declare(strict_types=1);

namespace Saola\Compiler\Template;

use Saola\Compiler\Support\Re;

/**
 * Kiểm tra cấu trúc thẻ component được import — dùng chung cho Blade và JS.
 *
 * Hai đích phân giải thẻ import ĐỘC LẬP nhau. Nguồn hỏng phải fail TRƯỚC khi
 * bên nào viết lại thẻ; nếu không, Blade có thể giữ lại một thẻ đóng lơ lửng
 * trong khi cây AST phía JS lặng lẽ bỏ qua nó — ra hai cây DOM khác nhau.
 *
 * HTML gốc cố ý để cho pipeline HTML/AST lo. Thẻ import thì nghiêm ngặt vì
 * chúng trở thành ranh giới vòng đời component.
 *
 * Port từ compiler/src/common/template_structure.py.
 */
final class TemplateStructure
{
    private const TAG_NAME = '/\G[A-Za-z][\w-]*/';

    /** Nội dung bên trong các thẻ này là văn bản thô, không phải markup. */
    private const RAWTEXT_TAGS = ['script', 'style', 'textarea', 'title'];

    private function __construct()
    {
    }

    /**
     * @param array<string, string> $imports thẻ → đường dẫn
     *
     * @throws TemplateStructureError
     */
    public static function validate(string $source, array $imports): void
    {
        if ($imports === []) {
            return;
        }

        /** @var list<array{string, int}> $stack [tên thẻ, vị trí mở] */
        $stack = [];
        $pos = 0;
        $length = strlen($source);

        while ($pos < $length) {
            if (str_starts_with(substr($source, $pos, 4), '{{--')) {
                $end = strpos($source, '--}}', $pos + 4);
                $pos = $end === false ? $length : $end + 4;
                continue;
            }

            if (str_starts_with(substr($source, $pos, 4), '<!--')) {
                $end = strpos($source, '-->', $pos + 4);
                $pos = $end === false ? $length : $end + 3;
                continue;
            }

            if ($source[$pos] !== '<') {
                $pos++;
                continue;
            }

            $cursor = $pos + 1;
            $isClosing = false;

            if ($cursor < $length && $source[$cursor] === '/') {
                $isClosing = true;
                $cursor++;
            }

            while ($cursor < $length && ctype_space($source[$cursor])) {
                $cursor++;
            }

            // \G neo vào đúng $cursor — tương đương regex.match(source, cursor)
            if (! Re::match(self::TAG_NAME, $source, $m, PREG_OFFSET_CAPTURE, $cursor)) {
                $pos++;
                continue;
            }

            $tagName = $m[0][0];
            $nameEnd = $cursor + strlen($tagName);
            $tagEnd = self::scanTagEnd($source, $nameEnd);
            $tagSource = substr($source, $pos, $tagEnd - $pos);

            // Không soi văn bản giống template bên trong thẻ raw/RCDATA gốc
            if (! $isClosing && in_array(strtolower($tagName), self::RAWTEXT_TAGS, true)) {
                $rest = substr($source, $tagEnd);
                $closePattern = '/<\/\s*' . preg_quote($tagName, '/') . '\s*>/i';

                if (Re::match($closePattern, $rest, $close, PREG_OFFSET_CAPTURE)) {
                    $pos = $tagEnd + $close[0][1] + strlen($close[0][0]);
                } else {
                    $pos = $length;
                }

                continue;
            }

            if (! isset($imports[$tagName])) {
                $pos = $tagEnd;
                continue;
            }

            if ($isClosing) {
                if ($stack === []) {
                    throw self::error(
                        $source,
                        $pos,
                        "Unexpected closing component tag </{$tagName}>; "
                        . "<{$tagName} /> is already self-closing",
                    );
                }

                [$openName, $openPos] = $stack[count($stack) - 1];

                if ($openName !== $tagName) {
                    [$openLine, $openColumn] = self::location($source, $openPos);

                    throw self::error(
                        $source,
                        $pos,
                        "Mismatched closing component tag </{$tagName}>; expected "
                        . "</{$openName}> for <{$openName}> opened at line "
                        . "{$openLine}, column {$openColumn}",
                    );
                }

                array_pop($stack);
            } elseif (! Re::match('/\/\s*>$/', $tagSource)) {
                $stack[] = [$tagName, $pos];
            }

            $pos = $tagEnd;
        }

        if ($stack !== []) {
            [$tagName, $openPos] = $stack[count($stack) - 1];

            throw self::error(
                $source,
                $openPos,
                "Unclosed component tag <{$tagName}>; expected </{$tagName}>",
            );
        }
    }

    /** Vị trí ngay sau dấu `>`, có tôn trọng thuộc tính nằm trong nháy. */
    private static function scanTagEnd(string $source, int $start): int
    {
        $quote = null;
        $escaped = false;
        $length = strlen($source);

        for ($pos = $start; $pos < $length; $pos++) {
            $char = $source[$pos];

            if ($quote !== null) {
                if ($escaped) {
                    $escaped = false;
                } elseif ($char === '\\') {
                    $escaped = true;
                } elseif ($char === $quote) {
                    $quote = null;
                }

                continue;
            }

            if ($char === "'" || $char === '"') {
                $quote = $char;
                continue;
            }

            if ($char === '>') {
                return $pos + 1;
            }
        }

        return $length;
    }

    /**
     * Dòng và CỘT của một vị trí.
     *
     * Cột tính theo CODEPOINT chứ không theo byte: thông báo lỗi phải trùng với
     * bản Python, mà Python đánh chỉ số theo codepoint. Dòng thì đếm '\n' nên
     * byte hay codepoint đều như nhau.
     *
     * @return array{int, int}
     */
    private static function location(string $source, int $index): array
    {
        $line = substr_count(substr($source, 0, $index), "\n") + 1;
        $lineStart = strrpos(substr($source, 0, $index), "\n");
        $lineStart = $lineStart === false ? 0 : $lineStart + 1;

        $column = mb_strlen(substr($source, $lineStart, $index - $lineStart), 'UTF-8') + 1;

        return [$line, $column];
    }

    private static function error(string $source, int $index, string $message): TemplateStructureError
    {
        [$line, $column] = self::location($source, $index);

        return new TemplateStructureError("{$message} at line {$line}, column {$column}.");
    }
}
